pallets/orml-tokens-extension/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+#![deny(warnings)]
+#![cfg_attr(test, feature(proc_macro_hygiene))]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+extern crate mocktopus;
+
+#[cfg(test)]
+use mocktopus::macros::mockable;
+use orml_traits::MultiCurrency;
+use sp_std::{convert::TryInto, prelude::*, vec};
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+#[cfg(test)]
+mod mock;
+
+pub mod default_weights;
+
+#[cfg(test)]
+mod tests;
+
+mod ext;
+
+mod types;
+
+pub use pallet::*;
+
+pub(crate) type BalanceOf<T> = <<T as orml_currencies::Config>::MultiCurrency as MultiCurrency<
+	<T as frame_system::Config>::AccountId,
+>>::Balance;
+
+pub(crate) type CurrencyOf<T> = <<T as orml_currencies::Config>::MultiCurrency as MultiCurrency<
+	<T as frame_system::Config>::AccountId,
+>>::CurrencyId;
+
+pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use crate::types::CurrencyDetails;
+	use frame_support::{pallet_prelude::*, transactional};
+	use frame_system::{ensure_root, ensure_signed, pallet_prelude::OriginFor};
+
+	pub use default_weights::WeightInfo;
+
+	/// ## Configuration
+	/// The pallet's configuration trait.
+	#[pallet::config]
+	pub trait Config: frame_system::Config + orml_tokens::Config + orml_currencies::Config {
+		/// The overarching event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Weight information for the extrinsics in this module.
+		type WeightInfo: WeightInfo;
+
+		/// Type that allows for checking if currency type is ownable by users
+		type CurrencyIdChecker: CurrencyIdCheck<CurrencyId = CurrencyOf<Self>>;
+
+		/// TODO needs to be conditionaly compiled
+		#[cfg(feature = "runtime-benchmarks")]
+		type GetTestCurrency: Get<CurrencyOf<Self>>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn currency_details)]
+	pub type CurrencyData<T: Config> =
+		StorageMap<_, Blake2_128Concat, CurrencyOf<T>, CurrencyDetails<AccountIdOf<T>>>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Some currency was issued.
+		Mint { currency_id: CurrencyOf<T>, to: AccountIdOf<T>, amount: BalanceOf<T> },
+		/// Some currency was burned.
+		Burned { currency_id: CurrencyOf<T>, from: AccountIdOf<T>, amount: BalanceOf<T> },
+		/// Some currency class was created.
+		Created { currency_id: CurrencyOf<T>, creator: AccountIdOf<T>, owner: AccountIdOf<T> },
+		/// Some currency data was destroyed
+		Destroyed { currency_id: CurrencyOf<T> },
```

Comment:
Maybe we should remove this event until we actually add the destruction of assets.